File: CIS194_intro_haskell/2013_spring/hw1/hanoi.ml
type peg = int list
type pegName = A | B | C
type step = Ab | Ac | Ba | Bc | Ca | Cb
type steps = step list
type state = State of (peg * peg * peg * steps)
exception Invalid_step of (pegName * pegName)
exception Invalid_place of (peg * peg)


let init (size : int) : state =
  let rec _seq (next : int) : int list =
    if next == size then [next]
    else next :: _seq (next + 1)
  in
  if size < 1 then State ([], [], [], [])
  else State (seq size, [], [], [])


let place (fromPeg : peg) (toPeg : peg) : (peg * peg) =
  match (fromPeg, toPeg) with
  | (fhead :: ftail, []) -> (ftail, [fhead])
  | (fhead :: ftail, thead :: _) when fhead < thead -> (ftail, fhead :: toPeg)
  | _ -> raise (Invalid_place (fromPeg, toPeg))


let step (fromPeg : pegName) (toPeg : pegName) (state : state) : state =
  let State (a, b, c, steps) = state in
  match (fromPeg, toPeg) with
  | (A, B) -> let (f, t) = place a b in State (f, t, c, Ab :: steps)
  | (A, C) -> let (f, t) = place a c in State (f, b, t, Ac :: steps)
  | (B, A) -> let (f, t) = place b a in State (t, f, c, Ba :: steps)
  | (B, C) -> let (f, t) = place b c in State (a, f, t, Bc :: steps)
  | (C, A) -> let (f, t) = place c a in State (t, b, f, Ca :: steps)
  | (C, B) -> let (f, t) = place c b in State (a, t, f, Cb :: steps)
  | _ -> raise (Invalid_step (fromPeg, toPeg))


let rec move (numDisks : int) (fromPeg : pegName) (toPeg : pegName) (tempPeg : pegName) (state : state) : state =
  match numDisks with
  | 0 -> state
  | 1 -> step fromPeg toPeg state
  | 2 -> state |> step fromPeg tempPeg |> step fromPeg toPeg |> step tempPeg toPeg
  | _ -> state
         |> move (numDisks - 1) fromPeg tempPeg toPeg
         |> step fromPeg toPeg
         |> move (numDisks - 1) tempPeg toPeg fromPeg


let solve (state : state) : state =
  let State (a0, _, _, _) = state in
  let State (a, b, c, steps) = move (List.length a0) A B C state in
  State (a, b, c, List.rev steps)

File: CIS194_intro_haskell/2013_spring/hw1/hanoi.scala
type Peg = List[Int]

sealed abstract class PegName
case object A extends PegName
case object B extends PegName
case object C extends PegName

sealed abstract class Step
case object Ab extends Step
case object Ac extends Step
case object Ba extends Step
case object Bc extends Step
case object Ca extends Step
case object Cb extends Step

type Steps = List[Step]

case class State(a: Peg, b: Peg, c: Peg, steps: Steps)

object Hanoi {

  def init(size: Int): State =
    new State((1 to size).toList, Nil, Nil, Nil)


  def solve(state: State): State = {
    val res = move(state.a.length, A, B, C, state)
    new State(res.a, res.b, res.c, res.steps.reverse)
  }


  def move(numDisks: Int, fromPeg: PegName, toPeg: PegName, tempPeg: PegName, state: State): State =
    numDisks match {
      case 0 => state
      case 1 => step(fromPeg, toPeg, state)
      case 2 =>
        step(tempPeg, toPeg,
          step(fromPeg, toPeg,
            step(fromPeg, tempPeg, state)))
      case _ =>
        val n = numDisks - 1
        move(n, tempPeg, toPeg, fromPeg,
          step(fromPeg, toPeg,
            move(n, fromPeg, tempPeg, toPeg, state)))
    }


  def step(fromPeg: PegName, toPeg: PegName, state: State): State =
    (fromPeg, toPeg) match {
      case (A, B) => val (f, t) = place(state.a, state.b); new State(f, t, state.c, Ab :: state.steps)
      case (A, C) => val (f, t) = place(state.a, state.c); new State(f, state.b, t, Ac :: state.steps)
      case (B, A) => val (f, t) = place(state.b, state.a); new State(t, f, state.c, Ba :: state.steps)
      case (B, C) => val (f, t) = place(state.b, state.c); new State(state.a, f, t, Bc :: state.steps)
      case (C, A) => val (f, t) = place(state.c, state.a); new State(t, state.b, f, Ca :: state.steps)
      case (C, B) => val (f, t) = place(state.c, state.b); new State(state.a, t, f, Cb :: state.steps)
      case _ => throw new IllegalArgumentException()
    }


  def place(fromPeg: Peg, toPeg: Peg): (Peg, Peg) =
    (fromPeg, toPeg) match {
      case (fhead :: ftail, Nil) => (ftail, List(fhead))
      case (fhead :: ftail, thead :: _) if fhead < thead => (ftail, fhead :: toPeg)
      case _ => throw new IllegalArgumentException()
    }
}
